### ISSUE TITLE
Fast Remote Memory  carlosgaldino

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6956]
Fast Remote Memory &middot; carlosgaldino<br>
<br>
https://ift.tt/3g1M705<br>
<br>
Carlos Galdino<br>
<br>
<div>
<div>
<div>
<span>26 Jul 2016</span>
<p>FaRM is a main memory distributed computing platform that provides distributed transactions with strict serializability, high performance, durability, and high availability.</p>
<p>To scale out, FaRM distributes objects across machines in a data center and also allows transactions to span any number of machines. To reduce CPU overhead it uses one-sided RDMA (Remote Direct Memory Access) operations.</p>
<p>Everything is stored in DRAM and it provides durability by attaching batteries to power supply units and writing the contents of DRAM to SSD when the power fails.</p>
<h2>Programming model and architecture</h2>
<p>FaRM provides the abstraction of a global shared address space that spans machines in a cluster. Each machine runs application threads and stores objects in the address space. The FaRM API provides transparent access to local and remote objects within transactions. Transactions can be started at any time and the coordinator will be whoever started it. The thread that started the transaction can execute arbitrary logic as well as read, write, allocate, and free objects. At the end of execution, the thread invokes FaRM to commit the transaction.</p>
<p>FaRM uses optimistic concurrency control. The operations are buffered locally and are only made visible to other transactions on successful commit.</p>
<p>FaRM provides strict serializability of all successfully committed transactions. Individual reads are atomic, reading only committed data, successive reads of the same object return the same data, and reading objects written by the transaction return the latest value written. Reads of different objects are not atomic but the transaction will not commit, thus the strict serializability property still holds.</p>
<p>Other features provided by the FaRM API is lock-free reads, which are optimized single-object read only transactions, and locality hints, enabling programmers to co-locate related objects on the same set of machines.</p>
<p>Each machine runs FaRM in a user process with a kernel thread pinned to each hardware thread. Each kernel thread runs an event loop that executes application code and polls the RDMA completion queues.</p>
<p>The configuration can change over time as machines fail or are added to the cluster. FaRM represent the configuration via a tuple, <code>{i, S, F, CM}</code>, where:</p>
<ul>
<li>
<code>i</code>: unique, monotonically increasing 64-bit configuration identifier;</li>
<li>
<code>S</code>: set of machines in the configuration;</li>
<li>
<code>F</code>: mapping from machines to failure domains that are expected to fail independently (e.g., different racks);</li>
<li>
<code>CM</code>: configuration manager. <code>CM &isin; S</code>.</li>
</ul>
<p>ZooKeeper is used as the coordination service to ensure that machines agree on the current configuration and store it. Lease management, failure detection and coordination recovery do not use ZooKeeper.</p>
<p>Regions of 2 GB are used as the global address space in FaRM and each region is replicated in <code>f + 1</code> machines, where <code>f</code> is the desired fault tolerance. Objects are always read from the primary copy of the containing region, and it uses local memory access if the region is on the local machine, otherwise one-sided RDMA read is used.</p>
<p>The <code>CM</code> is contacted when a machine wants to allocate a new region. The <code>CM</code> assigns a region identifier from a monotonically increasing counter and will select replicas for the region. Replica selection takes into account the number of regions stored on each machine, the available storage, and each replica is placed in a different failure domain. The region might be co-located with a target region if the application specifies a locality constraint. A two-phase protocol is used to coordinate the allocation in all replicas. The <code>CM</code> sends a prepare message to all replicas and they report if they allocated the region successfully. If all reply with success the <code>CM</code> sends a commit message to all of them.</p>
<p>Each machine also stores ring buffers that implement FIFO queues. The queues are used either as transaction logs or message queues.</p>
<p>The figure below shows the FaRM architecture:</p>
<div>
<div>
<pre><code>                   &#9484;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9488;
                   &#9474;                                  &#9474;
&#9484; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472;&#9474;&#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9488;     lease
                   &#9474;                Machine A      renewals
&#9474;                  &#9660;        CPU               &#9474;       &#9474;            &#9484; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472;
           &#9484;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9488;                          &#9474;              Machine D  &#9474;
&#9474;       &#9484;&#9472;&#9472;&#9474;  Application  &#9474;&#9664;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9488;         &#9474;       &#9492;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9474;    (CM)
        &#9474;  &#9500;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9508;        &#9474;                               &#9472; &#9472; &#9472; &#9472; &#9472; &#9472; &#9496;
&#9474;       &#9474;  &#9474;     FARM      &#9474;        &#9474;         &#9474;
     local &#9492;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9496;        &#9474;
&#9474;    reads       &#9650;   &#9650;              &#9474;         &#9474;
        &#9474;        &#9474;   &#9492;&#9472;&#9472;&#9472;&#9472;&#9488;         &#9474;
&#9474;       &#9660;        &#9474;        &#9474;  NVRAM  &#9474;         &#9474;
   &#9484;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9516;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9516;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9516;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9488;
&#9474;  &#9474; Region &#9474; Tx log &#9474; Tx log &#9474; Msg queue &#9474;   &#9474;
   &#9492;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9524;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9524;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9524;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9496;
&#9474;       &#9650;        &#9650;          &#9650;        &#9650;        &#9474;                         &#9484;&#9472;&#9488;
        &#9474;        &#9474;          &#9474;        &#9474;                            &#9484;&#9472;&#9472;&#9472;&#9472;&#9472;&#9474; &#9474;&#9472;&#9472;&#9472;&#9472;&#9472;&#9488;
&#9492; &#9472; &#9472; &#9472; &#9532; &#9472; &#9472; &#9472; &#9472;&#9474;&#9472; &#9472; &#9472; &#9472; &#9472; &#9532; &#9472; &#9472; &#9472; &#9472;&#9474;&#9472; &#9472; &#9472; &#9472; &#9496;                   &#9474;     &#9492;&#9472;&#9496;     &#9474;
        &#9474;        &#9474;          &#9474;        &#9474;                           &#9484;&#9472;&#9488;           &#9484;&#9472;&#9488;
      remote  tx records    &#9474;      messages                      &#9474; &#9474;           &#9474; &#9474;
      reads      &#9474;          &#9474;        &#9474;                           &#9492;&#9472;&#9496;           &#9492;&#9472;&#9496;
        &#9474;        &#9474;      tx records   &#9474;                            &#9474;             &#9474;
        &#9474;        &#9474;          &#9474;        &#9474;                            &#9474; &#9484;&#9472;&#9488;     &#9484;&#9472;&#9488; &#9474;
        &#9474;        &#9474;          &#9474;        &#9474;                            &#9492;&#9472;&#9474; &#9474;&#9472;&#9472;&#9472;&#9472;&#9472;&#9474; &#9474;&#9472;&#9496;
        &#9474;        &#9474;          &#9474;        &#9474;                              &#9492;&#9472;&#9496;     &#9492;&#9472;&#9496;
      &#9484; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472;       &#9484; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472;                             Coordination
        Machine B  &#9474;        Machine C  &#9474;                              service
      &#9492; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472;       &#9492; &#9472; &#9472; &#9472; &#9472; &#9472; &#9472;                             (ZooKeeper)
</code></pre>
</div>
</div>
<h2>Distributed transactions and replication</h2>
<p>To improve performance FaRM integrates the transaction and replication protocols. Both data and transaction logs use primary-backup replication in non-volatile DRAM, and it uses unreplicated transaction coordinators that communicate directly with primaries and backups. As mentioned before, FaRM uses optimistic concurrency control with read validation.</p>
<p>The timeline for a FaRM transaction is shown below:</p>
<p><img src="http://img.carlosgaldino.com/SOSP15-paper227-alternate-final-version.pdf_2016-07-22_15-41-01.png"></p>
<p>During the execution phase, transactions use one-sided RDMA to read objects and they buffer the writes locally. Versions and addresses of all accessed objects are recorded by the coordinator. For primaries and backups on the same machine as the coordinator, object reads and writes to the log use local memory accesses rather than RDMA. At the end of the execution, FaRM attempts to commit the transaction by performing the following steps:</p>
<ol>
<li>
<em>Lock</em>: The coordinator writes a <code>LOCK</code> record to the log of each machine that is a primary for any written object. This record contains the versions and new values of all written objects on that primary, as well as the list of all regions with written objects. The primaries process the records by attempting to lock the objects at the specified versions using compare-and-swap, and send back a message reporting if all locks were successfully taken. If any object version changed since it was read by the transaction, or if the object is currently locked by another transaction then the locking fails and the coordinator aborts the transaction. In this case, the coordinator writes an abort record to all primaries and returns an error to the application.</li>
<li>
<em>Validate</em>: All objects that were read but not written by the transaction are then validated by the coordinator by reading their versions from their primaries. The validation will fail and the transaction will abort if any object has changed.</li>
<li>
<em>Commit backups</em>: The coordinator writes a <code>COMMIT-BACKUP</code> record to the non-volatile logs at each backup and then waits for an ack from the NIC (Network Interface Controller) hardware without interrupting the backup&rsquo;s CPU. The <code>COMMIT-BACKUP</code> log record has the same payload as a <code>LOCK</code> record.</li>
<li>
<em>Commit primaries</em>: The coordinator writes a <code>COMMIT-PRIMARY</code> record to the log at each primary after all <code>COMMIT-BACKUP</code> writes have been acked. The coordinator then reports completion to the application after receiving at least one hardware ack for a <code>COMMIT-PRIMARY</code> record, or if the coordinator wrote one locally. The primaries process these records by updating the objects, incrementing their versions, and unlocking them, which exposes the writes committed by the transaction.</li>
<li>
<em>Truncate</em>: Both backups and primaries keep the records in their logs until they are truncated. Truncation is done lazily by the coordinator after it receives acks from all primaries. Backups apply the updates to their copies of the objects at truncation time.</li>
</ol>
<h3>Correctness</h3>
<p>The serialization of committed read-write transactions occurs when the locks are acquired, for committed read-only transactions it occurs at the point of their last read. This is because the versions of all read and written objects at the serialization point are the same as the versions seen during execution.</p>
<p>Serializability in FaRM is also <em>strict</em><sup><a href="#fn:1">1</a></sup>: the serialization point is always between the start of execution and the completion being reported to the application.</p>
<p>Even in case of failures FaRM ensures serializability. This is achieved by waiting acks for <em>all</em> backups before writing <code>COMMIT-PRIMARY</code>. Imagine that the coordinator does not receive an ack for a backup and writes <code>COMMIT-PRIMARY</code>. The application will receive a confirmation that the transaction was successful. If all replicas that stored the modification fail and the only backup surviving is the one that didn&rsquo;t acked because it never received <code>COMMIT-BACKUP</code>, this would result in losing the modification that was once confirmed.</p>
<p>For <code>COMMIT-PRIMARY</code> a single ack is enough for reporting back success to the application. This is true because FaRM assumes up to <code>f</code> failures for a <code>f + 1</code> shard. Then at least one machine will have a record attesting the commit. If not a single one <code>COMMIT-PRIMARY</code> was written and the primaries and backups failed, the <code>LOCK</code> records would survive but there wouldn&rsquo;t be any record attesting that validation succeeded.</p>
<h2>Failure recovery</h2>
<p>Durability and high availability are provided by replicating the data. FaRM assumes that machines can fail by crashing but can recover (fail-recovery) without losing the contents of non-volatile DRAM.</p>
<p>FaRM provides durability for all committed transactions even if the entire cluster fails or loses power: the committed state is recovered from regions and logs stored in non-volatile DRAM. It ensures durability even if at most <code>f</code> replicas per object lose the contents of non-volatile DRAM. FaRM can maintain availability with failures and network partitions provided that a majority of machines remain connected to each other and to a majority of replicas in the ZooKeeper service, and the partition contains at least one replica of each object.</p>
<h3>Failure detection</h3>
<p>FaRM uses leases to detect failures. Every machine other than the <code>CM</code> holds a lease at the <code>CM</code> and the <code>CM</code> holds a lease at every other machine. If any of these leases expire, failure recovery is then triggered.</p>
<p>FaRM uses dedicated queue pairs for leases to avoid having lease messages delayed in a shared queue behind other message types. It also uses a dedicated lease manager thread that runs at the highest user-space priority. This thread is not pinned to any hardware thread and it uses interrupts instead of polling to avoid starving critical OS tasks that are executed periodically on every hardware thread.</p>
<h3>Reconfiguration</h3>
<p>Reconfiguration happens when a failure occurs but there is a protocol that is followed to move a FaRM instance to the next configuration.</p>
<p>FaRM implements something they called <em>precise membership</em>. This guarantees that machines will only talk to other machines in the same configuration. Messages from machines outside the configuration are ignored.</p>
<p>The reconfiguration steps are as follows:</p>
<ol>
<li>
<em>Suspect</em>: If a lease for a machine at the <code>CM</code> expires, the <code>CM</code> suspects that a failure occurred for that machine and reconfiguration is initiated. The <code>CM</code> then starts blocking all external client requests. If a non-<code>CM</code> machine suspects that the CM failed due to a lease expiry, it will first ask a small number of &ldquo;backup <code>CM</code>s&rdquo; to initiate reconfiguration. If the configuration is unchanged after a timeout period it attempts the reconfiguration itself.</li>
<li>
<em>Probe</em>: An RDMA read is issued by the new <code>CM</code> to all machines in the configuration except the machine that is suspected. Any machine that the read fails is also suspected. To avoid having the new <code>CM</code> in a minority in case of a network partition, the new <code>CM</code> only proceeds with reconfiguration if it received responses from a majority of the probes.</li>
<li>
<em>Update configuration</em>: After receiving replies to the probes, the new <code>CM</code> attempts to update the configuration data stored on ZooKeeper.</li>
<li>
<em>Remap regions</em>: Regions that were previously mapped to failed machines are reassigned so that the number of replicas come back to <code>f + 1</code>. If a primary failed, a surviving backup is promoted to be the new primary. If the <code>CM</code> detects that regions lost all their replicas or if there is no space to re-replicate regions, it signals an error.</li>
<li>
<em>Send new configuration</em>: A <code>NEW-CONFIG</code> message is sent to all machines in the new configuration after the <code>CM</code> has remapped all regions. <code>NEW-CONFIG</code> also resets the lease protocol if the <code>CM</code> has changed.</li>
<li>
<em>Apply new configuration</em>: When a machine receives a <code>NEW-CONFIG</code> message with a configuration identifier that is greater than its own, it updates its current configuration identifier and its copy of the region mapping. It also allocates space to hold any new region replicas assigned to it. From this point on, the machine starts rejecting requests from machines that are not part of this new configuration and it does not issue new requests to those machines. Requests from external clients are also blocked. The machine will send a <code>NEW-CONFIG-ACK</code> message to <code>CM</code> as well.</li>
<li>
<em>Commit new configuration</em>: After receiving <code>NEW-CONFIG-ACK</code> messages from all machines in the configuration, the <code>CM</code> waits to ensure that any leases granted in previous configurations to machines not present in the new configuration have expired. Then it sends a <code>NEW-CONFIG-COMMIT</code> message to all configuration members. All members now unblock external requests and initiate transaction recovery.</li>
</ol>
<h3>Transaction state recovery</h3>
<p>After reconfiguration, FaRM recovers transaction state using the logs distributed across the replicas of objects modified by a transaction. This involves recovering the state both at the replicas of objects modified by the transaction and at the coordinator to determine the outcome of the transaction. This recovery is done by the following steps:</p>
<ol>
<li>
<em>Block access to recovering regions</em>: When the primary of a region fails, the backup is promoted to be the new primary. Access to this region is blocked until all transactions that updated it have been reflected at the new primary. Local access and RDMA references for the region in question are blocked until step 4 which is when the write locks have been acquired by all recovering transactions that updated the region.</li>
<li>
<em>Drain logs</em>: When trying to commit a transaction, the coordinators reply successfully to the application after receiving acks from NICs for <code>COMMIT-BACKUP</code> and <code>COMMIT-PRIMARY</code> messages. This just guarantees that the messages have been written to the logs but they might not have been processed yet. During transaction state recovery, the machines then drain the logs to ensure that all relevant records are processed during recovery. All records in the logs are processed in order after receiving a <code>NEW-CONFIG-COMMIT</code> message. At the end of it a <code>LastDrained</code> variable that stores the configuration identifier is updated. This <code>LastDrained</code> variable is used to reject log records for transactions with configuration identifiers that are less than or equal to <code>LastDrained</code>.</li>
<li>
<em>Find recovering transactions</em>: A recovering transaction is one whose commit phase spans configuration changes, and for which some replica of a written object, some primary of a read object, or the coordinator has changed due to reconfiguration. Only recovering transactions<sup><a href="#fn:2">2</a></sup> go through transaction recovery at primaries and backups. Records for a recovering may be distributed over the logs of different primaries and backups updated by the transaction. Each backup of a region sends a <code>NEED-RECOVERY</code> message to the primary with the configuration identifier, the region identifier, and the identifiers of recovering transactions that updated the region.</li>
<li>
<em>Lock recovery</em>: The primary of each region waits until the local machine logs have been drained and <code>NEED-RECOVERY</code> messages have been received from each backup, to build the complete set of recovering transactions that affect the region. The transactions are then sharded by identifier across its threads such that each thread <code>t</code> recovers the state of transactions with coordinator thread identifier <code>t</code>. In parallel, the threads from the primary fetch any transaction log records from backups that are not already stored locally and then lock any objects modified by the recovering transactions. The region is <em>active</em> when lock recovery is complete. Then local and remote coordinators can obtain local pointers and RDMA references, allowing them to read and commit updates to this region in parallel with subsequent recovery steps.</li>
<li>
<em>Replicate log records</em>: The threads in the primary replicates log records by sending backups the <code>REPLICATE-TX-STATE</code> message for any transactions that they are missing.</li>
<li>
<em>Vote</em>: The coordinator for a recovering transaction decides whether to commit or abort the transaction based on votes from each region updated by the transaction. The votes are sent by the primaries of each region.The threads in the primary send <code>RECOVERY-VOTE</code> messages to their peer threads in the coordinator for each recovering transaction that modified the region. The rules for voting are the following: * If any replica saw <code>COMMIT-PRIMARY</code> or <code>COMMIT-RECOVERY</code> then vote <em>commit-primary</em>. * If any replica saw <code>COMMIT-BACKUP</code> and did not see <code>ABORT-RECOVERY</code> then vote <em>commit-backup</em>. * If any replica saw <code>LOCK</code> and no <code>ABORT-RECOVERY</code> then vote <em>lock</em>. * Otherwise vote <em>abort</em>. If the primaries do not have any log records for the transaction they vote <em>truncated</em> if the transaction has already been truncated or <em>unknown</em> if it has not.</li>
<li>
<em>Decide</em>: The coordinator decides whether or not the transaction should commit. It decides to commit if it receives a <em>commit-primary</em> vote from any region. Otherwise, it waits for all regions to vote and commits if at least one region voted <em>commit-backup</em> and all other regions voted <em>lock</em>, <em>commit-backup</em>, or <em>truncated</em>. Otherwise it decides to abort. Then the coordinator sends <code>COMMIT-RECOVERY</code> or <code>ABORT-RECOVERY</code> to all participant replicas. A <code>TRUNCATE-RECOVERY</code> message is sent once the coordinator receives acks from all primaries and backups.</li>
</ol>
<h3>Recovering data</h3>
<p>To ensure that it can tolerate <code>f</code> replica failures in the future for a given region, FaRM must re-replicate data at new backups. Each machine sends a <code>REGIONS-ACTIVE</code> message to the <code>CM</code> when all regions for which it is primary become active. The <code>CM</code> after receiving all <code>REGIONS-ACTIVE</code> messages sends a <code>ALL-REGIONS-ACTIVE</code> message to all machines in the configuration. At this point, FaRM begins data recovery for new backups in parallel with foreground operations.</p>
<p>A new backup for a region initially has a freshly allocated and zeroed local region replica. The regions are divided across worker threads that recover them in parallel. Each thread issue one-sided RDMA operations to read a block at a time from the primary. Before being copied to the backup the recovered object is examined to see if a modification (by a new transaction) has been made while the backup was receiving the object.</p>
<h3>Recovering allocator state</h3>
<p>The FaRM allocator splits regions into blocks of 1 MB that are used as slabs for allocating small objects. Two pieces of meta-data re kept: block headers, which contain the object size, and slab free lists. The block headers are replicated to backups when a new block is allocated. Even if the primary fails they will be available at the new primary (which will be one of the backups). To avoid inconsistencies when the old primary fails while replicating the block header, the new primary sends all block headers to all backups immediately after receiving <code>NEW-CONFIG-COMMIT</code>.</p>
<p>The slab free lists are kept only at the primary but each object has a bit in its header that is set by an allocation and cleared by a free during transaction execution. After a failure, the free lists are recovered in the new primary by scanning the objects in the region.</p>
<h2>Conclusion</h2>
<p>In the paper&rsquo;s introduction the authors say that prior attempts to implement the abstraction of &ldquo;a single machine that never fails and that executes one transaction at a time in an order consistent with real time&rdquo; in a distributed system resulted in poor performance. They then cite several systems that sacrificed some characteristic; like strong consistency guarantees, transactions, etc; to have better performance.</p>
<p>That is what led them to build FaRM. The paper is a demonstration that new systems don&rsquo;t need to compromise those characteristics in order to have more performance. The ideas presented in the paper reflect this desire: reducing the number of messages exchanged, using one-sided RDMA to avoid using the CPU, storing the data in non-volatile DRAM, etc.</p>
<p>The performance results presented in the paper are interesting. FaRM achieves a peak throughput of 140 million TATP transactions per second on 90 machines, it recovers from a failure under 50 milliseconds, and it achieved 4.5 million TPC-C &ldquo;new order&rdquo; transactions per second. You can consult the paper for more numbers.</p>
<h2>References</h2>
<ul>
<li>Aleksandar Dragojevi&#263;, Dushyanth Narayanan, Edmund B Nightingale, Matthew Renzelmann, Alex Shamis, Anirudh Badam, and Miguel Castro. <em><strong>No compromises: distributed transactions with consistency, availability, and performance</strong></em>. In Proceedings of the 25th Symposium on Operating Systems Principles, October 04-07, 2015, Monterey, California.</li>
</ul>
<h2>Notes</h2>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via blog.carlosgaldino.com https://ift.tt/2X0ZAxE<br>
<br>
July 28, 2020 at 05:08PM
